fabric/src/main/java/com/discordsrv/fabric/requiredlinking/FabricRequiredLinkingModule.java

```diff
@@ -0,0 +1,297 @@
+        if(!enabled) return;
+        UUID playerUUID = serverPlayNetworkHandler.player.getUuid();
+
+        if(!loginsHandled.contains(playerUUID)) {
+            handleLogin(playerUUID, serverPlayNetworkHandler.player.getName().getString());
+        }
+
+        Component blockReason = frozen.get(playerUUID);
+        if (blockReason == null) {
+            return;
+        }
+
+        IPlayer srvPlayer = discordSRV.playerProvider().player(playerUUID);
+        if (srvPlayer == null) {
+            throw new IllegalStateException("Player not available: " + playerUUID);
+        }
+
+        srvPlayer.sendMessage(blockReason);
+    }
+
+    private void onPlayerQuit(ServerPlayNetworkHandler serverPlayNetworkHandler, MinecraftServer minecraftServer) {
+        if(!enabled) return;
+        UUID playerUUID = serverPlayNetworkHandler.player.getUuid();
+        loginsHandled.remove(playerUUID);
+        frozen.remove(playerUUID);
+    }
+
+    private void handleLogin(UUID playerUUID, String username) {
+        if (discordSRV.isShutdown()) {
+            return;
+        } else if (!discordSRV.isReady()) {
+            try {
+                discordSRV.waitForStatus(DiscordSRV.Status.CONNECTED);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+
+        ServerRequiredLinkingConfig config = config();
+        if (!config.enabled || config.action != ServerRequiredLinkingConfig.Action.FREEZE) {
+            return;
+        }
+
+        Component blockReason = getBlockReason(playerUUID, username, false).join();
+        if (blockReason != null) {
+            frozen.put(playerUUID, blockReason);
+        }
+    }
+
+    public static void onPlayerMove(ServerPlayerEntity player, PlayerMoveC2SPacket packet, CallbackInfo ci) {
+        if(instance == null || !instance.enabled) return;
+        Component freezeReason = instance.frozen.get(player.getUuid());
+        if (freezeReason == null) {
+            return;
+        }
+
+        double d = ServerPlayNetworkHandler.clampHorizontal(packet.getX(player.getX()));
+        double e = ServerPlayNetworkHandler.clampVertical(packet.getY(player.getY()));
+        double f = ServerPlayNetworkHandler.clampHorizontal(packet.getZ(player.getZ()));
+        double i = player.getX();
+        double j = player.getY();
+        double k = player.getZ();
+        double l = d - player.networkHandler.lastTickX;
+        double m = e - player.networkHandler.lastTickY;
+        double n = f - player.networkHandler.lastTickZ;
+        // If the player last position is the same. Don't move the player
+        if (l * l + m * m + n * n > 0.0001) {
+            player.networkHandler.requestTeleport(i, j, k, player.getYaw(), player.getPitch());
+            IPlayer iPlayer = instance.discordSRV.playerProvider().player(player);
+            iPlayer.sendMessage(freezeReason);
+
+        }
```

Comment:
This code is hard to understand. 

This also seems to block moving a tiny amount, which would lead to a huge amount of messages. 
The Bukkit version of this ignores decimals when comparing coordinates, it also let's the player fall down if they were in floating in the air for some reason